### ISSUE TITLE
"reportUnnessaryComparison", `bool` and `Literal[0, 1]`

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pyright#4861]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
Equality check between a `bool` value and an `int` literal (including `0` and `1`) is considered unnecessary, therefore "reportUnnecessaryComparison" is triggered.

**To Reproduce**
See "Screenshots or Code".

**Expected behavior**
The code may be sub-optimal, but I think "reportUnnecessaryComparison" should not take replace here.

**Screenshots or Code**
```python
# pyright: reportUnnessaryComparison=true

from typing import Literal

def f(flag: bool, zero_or_one: Literal[0, 1]):
    if flag == zero_or_one:
        ...
    else:
        ...
```
![image](https://user-images.githubusercontent.com/15389744/228479584-7ded50ee-7c63-494b-a5d9-68653f09da0b.png)

**VS Code extension or command-line**
Pylance v2023.3.31 Pre-Release

**Additional context**
None.

**Comment by erictraut** [comment]
The `reportUnnecessaryComparison` check works by looking for non-overlapping types in the two operands. When I implemented it, I knew there would be potential for false positives. That's why it's a strict-mode check only. 

Your choices are:
1. Disable the check
2. Make your code compatible with the assumptions of the check

```python
if flag == bool(zero_or_one): ...
```
